File: solutions/cpp/unique_binary_trees_all.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

#include "./Binary_tree_prototype.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::make_unique;
using std::random_device;
using std::stoi;
using std::uniform_int_distribution;
using std::vector;

unique_ptr<BinaryTreeNode<int>> Clone(const unique_ptr<BinaryTreeNode<int>>&);

// @include
vector<unique_ptr<BinaryTreeNode<int>>> GenerateAllBinaryTrees(
    int num_nodes) {
  vector<unique_ptr<BinaryTreeNode<int>>> result;
  if (num_nodes == 0) {  // Empty tree, add as an nullptr.
    result.emplace_back(nullptr);
  }

  for (int num_left_tree_nodes = 0; num_left_tree_nodes < num_nodes;
       ++num_left_tree_nodes) {
    int num_right_tree_nodes = num_nodes - 1 - num_left_tree_nodes;
    auto left_subtrees = GenerateAllBinaryTrees(num_left_tree_nodes);
    auto right_subtrees = GenerateAllBinaryTrees(num_right_tree_nodes);
    // Generates all combinations of left_subtrees and right_subtrees.
    for (auto& left : left_subtrees) {
      for (auto& right : right_subtrees) {
        result.emplace_back(make_unique<BinaryTreeNode<int>>(
            BinaryTreeNode<int>{0, Clone(left), Clone(right)}));
      }
    }
  }
  return result;
}

unique_ptr<BinaryTreeNode<int>> Clone(
    const unique_ptr<BinaryTreeNode<int>>& tree) {
  return tree ? make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{
                    0, Clone(tree->left), Clone(tree->right)})
              : nullptr;
}
// @exclude

void SmallTest() {
  assert(GenerateAllBinaryTrees(1).size() == 1);
  assert(GenerateAllBinaryTrees(2).size() == 2);
  assert(GenerateAllBinaryTrees(3).size() == 5);
  assert(GenerateAllBinaryTrees(4).size() == 14);
  assert(GenerateAllBinaryTrees(5).size() == 42);
  assert(GenerateAllBinaryTrees(10).size() == 16796);
}

int main(int argc, char** argv) {
  SmallTest();
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = stoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 10);
    n = dis(gen);
  }
  cout << "n = " << n << endl;
  GenerateAllBinaryTrees(n);
  return 0;
}

File: static/python/unique_binary_trees_all.py
import sys
import random
from binary_tree_prototype import BinaryTreeNode


# @include
def generate_all_binary_trees(num_nodes):
    if num_nodes == 0:  # Empty tree, add as a None.
        return [None]

    result = []
    for num_left_tree_nodes in range(num_nodes):
        num_right_tree_nodes = num_nodes - 1 - num_left_tree_nodes
        left_subtrees = generate_all_binary_trees(num_left_tree_nodes)
        right_subtrees = generate_all_binary_trees(num_right_tree_nodes)
        # Generates all combinations of left_subtrees and right_subtrees.
        result += [
            BinaryTreeNode(0, left, right)
            for left in left_subtrees for right in right_subtrees
        ]
    return result
# @exclude


def generate_all_binary_trees_pythonic(num_nodes):
    return [
        BinaryTreeNode(0, left, right)
        for num_left_tree_nodes in range(num_nodes)
        for left in generate_all_binary_trees(num_left_tree_nodes) for right in
        generate_all_binary_trees(num_nodes - 1 - num_left_tree_nodes)
    ] or [None]


def small_test():
    assert len(generate_all_binary_trees(1)) == len(
        generate_all_binary_trees_pythonic(1)) == 1
    assert len(generate_all_binary_trees(2)) == len(
        generate_all_binary_trees_pythonic(2)) == 2
    assert len(generate_all_binary_trees(3)) == len(
        generate_all_binary_trees_pythonic(3)) == 5
    assert len(generate_all_binary_trees(4)) == len(
        generate_all_binary_trees_pythonic(4)) == 14
    assert len(generate_all_binary_trees(5)) == len(
        generate_all_binary_trees_pythonic(5)) == 42
    assert len(generate_all_binary_trees(10)) == len(
        generate_all_binary_trees_pythonic(10)) == 16796


def main():
    small_test()
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10)
    print('n =', n)
    generate_all_binary_trees(n)


if __name__ == '__main__':
    main()
